File: Julia/textbook_track/chapter01/ba1m.jl
# Implement NumberToPattern

const NUCLEOTIDES = ('A', 'C', 'G', 'T')


function number_to_pattern(encoding, k)
    pattern = Char[]
    for _ in 1:k
        encoding, remainder = divrem(encoding, 4)
        push!(pattern, NUCLEOTIDES[remainder+1])
    end
    join(reverse(pattern))
end


function main()
    encoding = parse(Int64, readline())
    k = parse(Int64, readline())
    result = number_to_pattern(encoding, k)
    println(result)
end


main()

File: OCaml/src/textbook_track/chapter01/ba1m.ml
let nucleotides = ['A'; 'C'; 'G'; 'T']


let number_to_pattern (encoding : int) (k : int) : string =
    let rec loop pattern number l =
        if l = k then pattern |> List.to_seq |> String.of_seq
        else
            let number' = number / 4 in
            let remainder = number mod 4 in
            loop ((List.nth nucleotides remainder) :: pattern) number' (l + 1) in
    loop [] encoding 0


let () =
    let encoding = read_int () in
    let k = read_int () in
    let result = number_to_pattern encoding k in
    print_endline result
